<!DOCTYPE html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="theme-color" content="#2D2D2D" />
  
  <title>PEGG :: 🚀 Quickstart</title>
  
  <link rel="index" title="Index" href="genindex.html"/>

  <link rel="stylesheet" href="_static/css/nltk_theme.css"/>
  <link rel="stylesheet" href="_static/css/custom.css"/>

  <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
      <script type="text/javascript" src="_static/documentation_options.js"></script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
  

  <script src="https://email.tl.fortawesome.com/c/eJxNjUEOgyAQAF8jR7Kw6wIHDh7sP1Cw2mgxgmn6-3JsMqc5zEQfE8dkxOY1KKMUOI3ACFKRJpSW2AAp7ontYIaxI6i7XPJVwyeVfCQ550Os3jLrGSNOLgbdAy6s0PBk2TFNjEbsfq31LB0OnX407pJa5v2faRadwSW63mn5KuLyR9j2tgx3zecanl-55R_-jjPs"></script> 
</head>

<body>
  <div id="nltk-theme-container">
    <header>
      <div id="logo-container">
          
          <a href="index.html"><img src="_static/PEGG_notext.png"></a>
          
      </div>
      <div id="project-container">
        
        <h1>PEGG Documentation</h1>
        
      </div>

      <a id="menu-toggle" class="fa fa-bars" aria-hidden="true"></a>

      <script type="text/javascript">
        $("#menu-toggle").click(function() {
          $("#menu-toggle").toggleClass("toggled");
          $("#side-menu-container").slideToggle(300);
        });
      </script>
    </header>

    <div id="content-container">

      <div id="side-menu-container">

        <div id="search" role="search">
        <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
            <input type="text" name="q" placeholder="Search" />
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
        </form>
</div>

        <div id="side-menu" role="navigation">
          
  
    
  
  
    <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">🚀 Quickstart</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#required-input-files">Required Input Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#reference-file-formatting">Reference File Formatting</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#input-mutations">(1) Input mutations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#reference-genome">(2) Reference Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#on-off-target-scores">(3) On/Off Target scores</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#using-pegg">Using PEGG</a></li>
<li class="toctree-l2"><a class="reference internal" href="#jupyter-notebook-tutorial">Jupyter Notebook Tutorial</a></li>
<li class="toctree-l2"><a class="reference internal" href="#a-note-on-ram">A Note on RAM</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="PEGG.html">Complete Function Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="about.html">About PEGG</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">src</a><ul>
<li class="toctree-l2"><a class="reference internal" href="PEGG.html">Complete Function Documentation</a></li>
</ul>
</li>
</ul>

  

        </div>

        
      </div>

      <div id="main-content-container">
        <div id="main-content" role="main">
          
  <section id="quickstart">
<h1>🚀 Quickstart<a class="headerlink" href="#quickstart" title="Permalink to this headline">¶</a></h1>
<section id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p>PEGG is available through the python package index. To install, use pip:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">pegg</span>
</pre></div>
</div>
</section>
<section id="required-input-files">
<h2>Required Input Files<a class="headerlink" href="#required-input-files" title="Permalink to this headline">¶</a></h2>
<p>In order to use PEGG, you need at minimum 2 input files:</p>
<ol class="arabic simple">
<li><p>A dataframe containing a list of input mutations</p></li>
<li><p>A reference genome to work from.</p></li>
</ol>
<p>An optional, but highly reccomended input file for use with the reference genome build GrCh37 is:</p>
<ol class="arabic simple" start="3">
<li><p>chrom_dict, which provides on/off target scores for each protospacer occuring in exonic regions of GrCh37.</p></li>
</ol>
<p>Reference files for (2) and (3), as well as a sample mutation dataset (1), are available at the following link:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.dropbox.com/sh/h6fdvpv3tyny27q/AADYVOkJe12XZiD4pf3_WXuga?dl=0">Reference Files Dropbox Link</a></p></li>
</ul>
</section>
<section id="reference-file-formatting">
<h2>Reference File Formatting<a class="headerlink" href="#reference-file-formatting" title="Permalink to this headline">¶</a></h2>
<p>PEGG requires that the reference files are in a specific format. This should be the only painful part of setting up your environment
to allow PEGG to generate pegRNAs.</p>
<section id="input-mutations">
<h3>(1) Input mutations<a class="headerlink" href="#input-mutations" title="Permalink to this headline">¶</a></h3>
<p>First let’s load in a dataframe containing a list of input mutations (1):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pegg</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">gzip</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="c1">#pandas, gzip, and Bio.SeqIO are required for parsing the reference files</span>
<span class="c1">#they can be easily downloaded using &#39;pip install _____&#39;</span>

<span class="c1">#------(1) loading in input mutatations-------------</span>
<span class="n">filepath</span> <span class="o">=</span> <span class="s1">&#39;.../2020-06-16-MSK-IMPACT_EDITED.txt&#39;</span>
<span class="n">mutant_input</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The above codeblock shows how to load in the sample mutation input file available in the <a class="reference external" href="https://www.dropbox.com/sh/h6fdvpv3tyny27q/AADYVOkJe12XZiD4pf3_WXuga?dl=0">dropbox of reference files</a>.
To use <strong>your own mutation input</strong>, the input mutations follow the format shown in the dataframe below:</p>
<img alt="_images/mutant_input.png" src="_images/mutant_input.png" />
<p>Namely, 7 columns are required, providing information about the input mutations relative to the provided reference genome.
The input mutation dataframe <strong>must match the column naming scheme shown above</strong>:</p>
<ol class="arabic simple">
<li><p>Hugo_Symbol: the name of the gene within which the mutations falls. For non-coding locations, include some identifer.</p></li>
<li><p>Chromosome: which chromosome the mutation falls in. Use integers or ‘X’ and ‘Y’.</p></li>
<li><p>Start_Position: start position of mutation in the reference genome (mutations should only be reported on the + strand)</p></li>
<li><p>End_Position: end position of mutation in the reference genome</p></li>
<li><p>Variant_Type: what type of mutation is it. Options: “SNP”, “ONP”, “INS”, “DEL”</p></li>
<li><p>Reference_Allele: what is the reference allele. For insertions, this can be set to “-“.</p></li>
<li><p>Tumor_Seq_Allele2: what is the mutant allele (i.e. what is the mutation sequence). For deletions, this can be set to “-“.</p></li>
</ol>
</section>
<section id="reference-genome">
<h3>(2) Reference Genome<a class="headerlink" href="#reference-genome" title="Permalink to this headline">¶</a></h3>
<p>And now let’s load in the reference genome. I suggest using the codeblock below if you’re using GrCh37 as a reference file.
Otherwise, the format needs to be as follows:</p>
<ul>
<li><p><strong>records</strong> = a list containing the chromosome sequences (preferably parsed by SeqIO).</p></li>
<li><p><strong>index_list</strong> = a list containing the indeces in <strong>records</strong> corresponding to chromosome 1-22, and chrX and chrY.</p>
<blockquote>
<div><ul class="simple">
<li><p>For example, <strong>records[index_list[0]] should refer to chromosome 1</strong> (numbering starts at 0, not 1 in index_list due to python conventions)</p></li>
<li><p>And <strong>records[index_list[1]] should refer to chromosome 2</strong>, etc.</p></li>
<li><p>For the X and Y chromosome, records[index_list[22]] should refer to chrX, and records[index_list[23]] should refer to chrY</p></li>
</ul>
</div></blockquote>
</li>
</ul>
<p>In other words, <strong>records</strong> is a list containing the chromosome sequences, but likely also other unwanted sequences (alternate/unlocalized sequences),
while <strong>index_list</strong> provides the indeces needed to place the chromosome sequences in records in order from chromosome 1-22, and chrX and chrY.</p>
<ul class="simple">
<li><p>An alternate way to format these input files is to have <strong>records</strong> list have the chromosome sequences in order from 1-22, chrX, and chrY, and set <strong>index_list</strong> = list(range(24))</p></li>
</ul>
<p>See the following codeblock for how to set up <strong>records</strong> and <strong>index_list</strong> using the <a class="reference external" href="https://www.dropbox.com/sh/h6fdvpv3tyny27q/AADYVOkJe12XZiD4pf3_WXuga?dl=0">provided reference genome file, GrCh37</a>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#------(2) loading in reference genome GrCh37-------------</span>
<span class="n">file</span> <span class="o">=</span> <span class="s1">&#39;.../GRCh37/ncbi-genomes-2022-03-17/GCF_000001405.25_GRCh37.p13_genomic.fna.gz&#39;</span>

<span class="c1">#using gzip to load in the file...</span>
<span class="k">with</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s2">&quot;rt&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
   <span class="n">records</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">))</span> <span class="c1">#about 4 Gb in  memory</span>
   <span class="c1">#records = list that contains sequences split up by chromosome</span>

<span class="c1">#filtering out alternative sequences to only select consensus chromosome sequences</span>
<span class="n">wrong</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;alternate&quot;</span><span class="p">,</span> <span class="s2">&quot;unplaced&quot;</span><span class="p">,</span> <span class="s2">&quot;unlocalized&quot;</span><span class="p">,</span> <span class="s2">&quot;patch&quot;</span><span class="p">]</span>
<span class="n">badlist</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">wrong</span><span class="p">:</span>
   <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
      <span class="n">ii</span> <span class="o">=</span> <span class="n">i</span><span class="o">.</span><span class="n">description</span>
      <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">ii</span><span class="p">:</span>
            <span class="n">badlist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ii</span><span class="p">)</span>

<span class="c1">#creating an index for</span>
<span class="n">index_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">records</span><span class="p">):</span>
   <span class="n">ii</span> <span class="o">=</span> <span class="n">i</span><span class="o">.</span><span class="n">description</span>
   <span class="k">if</span> <span class="n">ii</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">badlist</span><span class="p">:</span>
      <span class="n">index_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="on-off-target-scores">
<h3>(3) On/Off Target scores<a class="headerlink" href="#on-off-target-scores" title="Permalink to this headline">¶</a></h3>
<p>The last reference file (3) is chrom_dict, which provides on/off target scores for protospacers in exonic regions of GrCh37.
These scores are used in the ranking and filtration of pegRNAs. Curently this is only available for GrCh37, but a chrom_dict reference file for
GrCh38 will be provided shortly and available in the <a class="reference external" href="https://www.dropbox.com/sh/h6fdvpv3tyny27q/AADYVOkJe12XZiD4pf3_WXuga?dl=0">dropbox of reference files</a>.</p>
<p>To import this reference file, use the following syntax:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#---------loading in on/off-target efficiencies of sgRNAs in exonic regions of GrCh37---------------------</span>
<span class="n">file</span> <span class="o">=</span> <span class="s1">&#39;.../chrom_dict.pickle&#39;</span>
<span class="n">chrom_dict</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_pickle</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
</pre></div>
</div>
<p>If you are using an alternative genome, or don’t wish to use chrom_dict, set it equal to ‘none’:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">chrom_dict</span> <span class="o">=</span> <span class="s1">&#39;none&#39;</span>
</pre></div>
</div>
</section>
</section>
<section id="using-pegg">
<h2>Using PEGG<a class="headerlink" href="#using-pegg" title="Permalink to this headline">¶</a></h2>
<p>Now that our reference files are loaded in, and PEGG is imported as a module, using PEGG is simple.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#specify which mutations within mutant_input to generate pegRNAs for</span>
<span class="c1">#here we&#39;re going to just generate pegRNAs for 2 mutations, corresponding to row 422804 and row 16 of mutant_input</span>
<span class="n">mut_idx_list</span> <span class="o">=</span> <span class="p">[</span><span class="mi">422804</span><span class="p">,</span> <span class="mi">16</span><span class="p">]</span> <span class="c1">#alternatively, we can generate pegRNAs for the entire dataset = list(range(len(mutant_input)))</span>
<span class="n">PAM</span> <span class="o">=</span> <span class="s1">&#39;NGG&#39;</span> <span class="c1">#specify PAM sequence</span>
<span class="n">guides_per_mut</span> <span class="o">=</span> <span class="mi">5</span>    <span class="c1">#specify how many pegRNAs to return for each mutation</span>
<span class="n">RTT_lengths</span> <span class="o">=</span> <span class="p">[</span><span class="mi">20</span><span class="p">,</span><span class="mi">25</span><span class="p">,</span><span class="mi">30</span><span class="p">]</span> <span class="c1">#specify RTT lengths and PBS lengths to generate pegRNAs for</span>
<span class="n">PBS_lengths</span> <span class="o">=</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">10</span><span class="p">]</span>
<span class="n">minus_seqs</span> <span class="o">=</span> <span class="n">pegg</span><span class="o">.</span><span class="n">minus_seq_generator</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">index_list</span><span class="p">)</span>

<span class="c1">#now generating the pegRNAs</span>
<span class="n">ranked_filtered_output</span> <span class="o">=</span> <span class="n">pegg</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">mutant_input</span><span class="p">,</span> <span class="n">mut_idx_list</span><span class="p">,</span> <span class="n">records</span><span class="p">,</span> <span class="n">index_list</span><span class="p">,</span> <span class="n">minus_seqs</span><span class="p">,</span> <span class="n">chrom_dict</span><span class="p">,</span> <span class="n">PAM</span><span class="p">,</span> <span class="n">RTT_lengths</span><span class="p">,</span> <span class="n">PBS_lengths</span><span class="p">,</span> <span class="n">guides_per_mut</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="jupyter-notebook-tutorial">
<h2>Jupyter Notebook Tutorial<a class="headerlink" href="#jupyter-notebook-tutorial" title="Permalink to this headline">¶</a></h2>
<p>A jupyter notebook version of the PEGG tutorial can be accessed at the following link:</p>
</section>
<section id="a-note-on-ram">
<h2>A Note on RAM<a class="headerlink" href="#a-note-on-ram" title="Permalink to this headline">¶</a></h2>
<p>Importing a reference genome into the local environment requires ~4 Gb of RAM. Chrom_dict is also a large file.
It’s reccomended to use a machine with <em>at least</em>  16 Gb of RAM, though more is preferable, when running pegg.</p>
</section>
</section>


        </div>
      </div>

    </div>

<footer>
    <div id="footer-info">
        <ul id="build-details">
            
                <li class="footer-element">
                    
                        <a href="_sources/quickstart.rst.txt" rel="nofollow"> source</a>
                    
                </li>
            

            

            
        </ul>

        
            <div id="copyright">
                &copy; 2022, Samuel Gould
            </div>
        

        <div id="credit">
            created with <a href="http://sphinx-doc.org/">Sphinx</a> and <a href="https://github.com/tomaarsen/nltk_theme">NLTK Theme</a>
        </div>
    </div>
</footer> 

</div>

</body>
</html>